lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -78,7 +78,10 @@ struct ExprVisitor {
 
   // Handle blocking and non-blocking assignments.
   Value visit(const slang::ast::AssignmentExpression &expr) {
-    auto lhs = context.convertExpression(expr.left());
+    auto *readOp = context.convertExpression(expr.left()).getDefiningOp();
+    auto lhs = dyn_cast<moore::ReadLValueOp>(readOp).getOperand();
+    readOp->erase();
+
```

Comment:
There is no way to estimate when to create `moore.read_lvalue`. So call `op->erase()` to remove it if it is created by `expr.left()`.

Reply:
I think this is a good idea!

I have a feeling that we are going to introduce a dedicated expression lowering for left-hand sides of assignments. This is done in a similar fashion in C/C++:
```c
int x, y;
x = y
```
would lower to something like the following pseudo-code:
```mlir
%x = alloca : ptr<int>
%y = alloca : ptr<int>
%0 = load %y  // y lowered as rvalue
store %x, %0  // x lowered as lvalue
```

If arrays or structs are involved, you'd have dedicated ops that allow you to convert a pointer/reference to the struct into a reference to the field you're interested in:
```c
struct { int a, b; } x, y;
x.b = y.a;
```
would lower to something like this:
```mlir
%x = alloca : ptr<struct<a: int, b: int>>
%y = alloca : ptr<struct<a: int, b: int>>
// x.b lowered as lvalue
%0 = struct_field_ref %x, "b" : ptr<struct<a: int, b: int>> -> ptr<int>
// y.a lowered as rvalue
%1 = load %y : ptr<struct<a: int, b: int>> -> struct<a: int, b: int>
%2 = struct_field %x, "a" : struct<a: int, b: int> -> int
// assignment
store %0, %2 : ptr<int>, int
```

These ops would allow us to reason about assignable lvalues that are just references to a storage location, and rvalues that are actual values. We would probably create two separate expression lowering functions, `context.convertLvalueExpression` and `context.convertRvalueExpression`: the former would always return a `ref<T>` type, and the latter would always return an actual value `T`.